spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/ReflectUtils.scala

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2016 The BigDL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.utils.intermediate
+
+import com.intel.analytics.bigdl._
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.serializer.ModuleSerializer._
+import scala.collection.mutable
+import scala.reflect.{ClassTag, ManifestFactory}
+import scala.reflect.runtime._
+
+object ReflectUtils {
+
+  private def getFieldNameAndValues(o: Object): mutable.HashMap[String, AnyRef] = {
+    val c = o.getClass
+    var fields = c.getDeclaredFields
+    val superFields = c.getSuperclass.getDeclaredFields
+    fields = fields ++ superFields
+
+    val values = new mutable.HashMap[String, AnyRef]()
+    fields.foreach(field => {
+      field.setAccessible(true)
+      values(field.getName) = field.get(o)
+    })
+    values
+  }
+
+  // create layer2 object form layer1
+  private def reflection(layer1: Object, layer2: Class[_],
+     tags: Array[ClassTag[_]], numerics: Array[TensorNumeric[_]]) : Object = {
+    val nameAndValues = getFieldNameAndValues(layer1)
+    val constructorMirror = getCostructorMirror(layer2)
+    val constructorFullParams = constructorMirror.symbol.paramss
+    val args = new Array[Object](constructorFullParams.map(_.size).sum)
+
+    val tagIter = tags.iterator
+    val numericIter = numerics.iterator
+    var i = 0
+    constructorFullParams.foreach(map => {
+      map.foreach(param => {
+        val name = param.name.decodedName.toString
+        val ptype = param.typeSignature
+        if (ptype <:< universe.typeOf[ClassTag[_]]||
+          ptype.typeSymbol == universe.typeOf[ClassTag[_]].typeSymbol) {
+          require(tagIter.hasNext, "If your module contains multiple class tags, " +
+            s"do you forget to override getClassTagNumerics method ${layer1}")
+          args(i) = tagIter.next()
+        } else if (ptype <:< universe.typeOf[TensorNumeric[_]]
+          || ptype.typeSymbol == universe.typeOf[TensorNumeric[_]].typeSymbol) {
+          args(i) = numericIter.next()
+        } else {
+          val value = nameAndValues.get(name).getOrElse(null)
+          args(i) = value
+        }
+        i += 1
+      })
+    })
+    constructorMirror.apply(args : _*).asInstanceOf[Object]
+  }
+
+  // create Module form IRElement
+  def reflectFromIR[T : ClassTag](layer: IRElement[T], cls: Class[_]) : Module[T] = {
```

Comment:
the class name already have reflect prefix, do you think we still need reflect in method?

Reply:
sorry, I cannot understand your meaning. Do you mean remove "reflectFromIR" here and do reflect in IRToBlas or IRToDnn?

Reply:
Sorry, i should have been more clear it. I just meant the naming here, the class name has "Refelect", so do we need `reflect` in method name ?  `ReflectionUtils.reflectFromIR` or `ReflectionUtils.fromIR` which is better?

Reply:
I think reflectFromIR  is more clear. After all, ReflectionUtils has many funcs.